source/event.hpp

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <any>
+#include <queue>
+
+#include "device.hpp"
+#include "flow.hpp"
+#include "packet.hpp"
+
+namespace sim {
+
+// Base class for event
+struct Event {
+    int time;
+    virtual ~Event();
+    virtual void operator()() = 0;
+};
+
+/**
+ * Generate a packet in the flow and enqueue it in
+ * the source node ingress buffer for processing.
+ * Schedule the next packet generation event.
+ */
+struct Generate : public Event {
+    Generate(Flow *a_flow);
```

Comment:
Maybe `packet_size` is needed to be passed as a parameter? Since we are trying to generate packets here
```suggestion
    Generate(Flow *a_flow, int a_packet_size);
```